odl/operator/discr_ops.py

```diff
@@ -0,0 +1,446 @@
+
+        for axis in range(ndim):
+            finite_diff(x_data, out=out[axis].asarray(), axis=axis,
+                        dx=dx[axis], edge_order=2, zero_padding=True)
+
+    @property
+    def adjoint(self):
+        """Return the adjoint operator given by the negative of the
+        `DiscreteDivergence` operator assuming implicit zero padding.
+
+        Note that the `space` argument of the `DiscreteDivergence` operator
+        is not the range but the domain of the `DiscreteGradient` operator.
+        """
+        return - DiscreteDivergence(self.domain)
+
+
+class DiscreteDivergence(Operator):
+    """Divergence operator applicable to `DiscreteLp` spaces with number of
+    dimensions ``n``.
+
+    Calls helper function `finite_diff` for each component of the input
+    product space vector. For the adjoint of the `DiscreteDivergence`
+    operator to match the negative `DiscreteGradient` operator implicit zero
+    padding is assumed.
+    """
+
+    def __init__(self, space):
+        """Initialize a `DiscreteDivergence` operator instance.
+
+        Zero padding is assumed for the negative `DiscreteGradient`
+        operator to match the adjoint of the `DiscreteDivergence` operator.
+
+        Parameters
+        ----------
+        space : `DiscreteLp`
+            The space of elements which the operator is acting on
+        """
+        if not isinstance(space, DiscreteLp):
+            raise TypeError('space {!r} is not a `DiscreteLp` '
+                            'instance.'.format(space))
+
+        self.space = space
+        super().__init__(domain=ProductSpace(space, space.grid.ndim),
+                         range=space, linear=True)
+
+    def _apply(self, x, out):
+        """Apply the divergence operator to ``x`` and store result in ``out``.
+
+        Parameters
+        ----------
+        x : ``domain`` element
+            `ProductSpaceVector` to which the divergence operator is applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import Rectangle, uniform_discr, FunctionSpace
+        >>> data = np.arange(5, dtype=float) * np.arange(1, 3).reshape(2,1)
+        >>> space = FunctionSpace(Rectangle([0,0], [2,1]))
+        >>> disc = uniform_discr(space, data.shape)
+        >>> div = DiscreteDivergence(disc)
+        >>> f = div.domain.element([data, data ** 2])
+        >>> div_f = div(f)
+        >>> print(div_f)
+        [[2.5, 11.0, 22.0, 33.0, -18.5],
+         [10.0, 39.5, 79.0, 118.5, -92.0]]
+        >>> g = div.range.element(data ** 3)
+        >>> adj = div.adjoint
+        >>> adj_g = adj(g)
+        >>> g.inner(div_f)
+        -3248.4
+        >>> f.inner(adj_g)
+        -3248.4
+        """
+
+        ndim = self.range.grid.ndim
+        dx = self.range.grid.stride
+
+        # TODO: can the following code be optimized?
+        tmp = np.zeros_like(out.asarray())
```

Comment:
Without calling `asarray()`
```python
tmp = np.zeros(out.shape, out.dtype, order=out.order)
```

I'd write this:

```python
arr = out.asarray()
tmp = np.zeros(out.shape, out.dtype, order=out.order)
for axis in range(ndim):
    arr += finite_diff(x[axis], out=tmp, axis=axis, dx=dx[axis],
                       edge_order=2, zero_padding=True)
out[:] = arr # self assignment in case asarray is a view, thus no overhead.
```

This is a pattern you could use in general.